### DISCUSSION TITLE
Windows - Fips provider is not loading

### DISCUSSION THREAD

**Discussion by Harish-Kumarr** [openssl/openssl#21678] [Q&A]
@t8m 
Openssl version used -  3.0.8

Using the below program, trying to load the fips provider & other providers but only for fips provider null value is showing.
Attaching the configuration.zip( contains openssl.cnf & fipsmodule.cnf) 
[configuration.zip](https://github.com/openssl/openssl/files/12258440/configuration.zip)

```
#include "stdafx.h"
#include <iostream>
#include <openssl/evp.h>
#include <openssl/provider.h>
int main()
{


	OSSL_PROVIDER *fips;
	OSSL_PROVIDER *base;
	OSSL_PROVIDER *legacy;
	OSSL_PROVIDER *deflt;

	if (!OSSL_LIB_CTX_load_config(NULL, "C:\\Program Files (x86)\\Common Files\\openSSL3.0.8\\ssl\\openssl.cnf"))
	{
		//printf("%d", &k);
		printf("Loaded config file Failed");
		getchar();
	}

	fips = OSSL_PROVIDER_load(NULL, "fips");
	if (fips != NULL) {
		printf("FIPS provider loaded\n");
		getchar();
	}
	else 
	{
		printf("Failed to load FIPS provider\n");
		getchar();
	}
	base = OSSL_PROVIDER_load(NULL, "base");
	if (base == NULL) {
		OSSL_PROVIDER_unload(fips);
		printf("Failed to load legacy provider\n");
	}
	legacy = OSSL_PROVIDER_load(NULL, "base");
	if (legacy == NULL) {
		OSSL_PROVIDER_unload(base);
		printf("Failed to load legacy provider\n");
	}

	deflt = OSSL_PROVIDER_load(NULL, "default");
	if (deflt == NULL) {
		printf("Failed to load Default provider\n");
		OSSL_PROVIDER_unload(legacy);
	}
	  return 0;
}
```

Attaching the error screenshort 
![Screenshot (8574)](https://github.com/openssl/openssl/assets/93762088/9d69bdf9-7eb2-4ef3-b2c9-ed83441f703b)
![Screenshot (8572)](https://github.com/openssl/openssl/assets/93762088/01056cdd-682e-4189-9936-523f8fdbff67)

**Comment by junaruga** [comment]
How about the result by the following command to list the providers?

```
$ openssl list -providers
```

In your `openssl.cnf`, the absolute path of the `fipsmodule.cnf` file is correct?

```
...
.include C:\\Program Files (x86)\\Common Files\\openSSL3.0.8\\ssl\\fipsmodule.cnf
...
```

Did you confirmed that your fips library is actually installed? I am not familiar with the Windows case. But in the Linux case, there is the `./Configure enable-fips` option to configure to enable to install the FIPS module. Did you set it?

In the Linux case, for this kind of issue, I used [strace](https://github.com/strace/strace) command to trace a system call's log, and checked if a openssl program was actually calling the fips module library. Is there an equivalent command to check this in Windows?

**Comment by paulidale** [comment]
Please print out the error stack so we've some idea what's going awry.
`ERR_print_errors_fp`

**Comment by slontis** [comment]
The most common things to go wrong are:
1) It cant find the library/or config path.. There are environment variables for
OPENSSL_CONF_INCLUDE
OPENSSL_MODULES
2) The FIPS self tests fail.
Any time the source changes the fipsmodule.cnf file needs to be updated to contain the correct module-mac

**Comment by t8m** [comment]
> 2\. Any time the source changes the fipsmodule.cnf file needs to be updated to contain the correct module-mac

Not just the source. Even a rebuild of the same sources will almost certainly produce a different binary - which has a different module-mac.
